flaky test: TestRateLimitedResolve

```
--- FAIL: TestRateLimitedResolve (10.10s)
    dns_resolver_test.go:1531: Timed out waiting for 3 calls after 40ms; got 2
    leakcheck.go:110: Leaked goroutine: goroutine 152 [sleep]:
        time.Sleep(0xf4240)
        	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/time.go:193 +0xcc
        google.golang.org/grpc/internal/resolver/dns.TestRateLimitedResolve.func3(0x40000922a0, 0x385f18, 0x40000da070)
        	/home/runner/work/grpc-go/grpc-go/internal/resolver/dns/dns_resolver_test.go:1517 +0x44
        created by google.golang.org/grpc/internal/resolver/dns.TestRateLimitedResolve
        	/home/runner/work/grpc-go/grpc-go/internal/resolver/dns/dns_resolver_test.go:1510 +0x2dc
FAIL
```

https://github.com/grpc/grpc-go/pull/4356/checks?check_run_id=2422242299

Have only seen this once on the newly added `arm64` tests

--- a/internal/resolver/dns/dns_resolver.go
+++ b/internal/resolver/dns/dns_resolver.go
@@ -47,7 +47,8 @@ var EnableSRVLookups = false
 
 var logger = grpclog.Component("dns")
 
-// Globals to stub out in tests.
+// Globals to stub out in tests. TODO: Perhaps these two can be combined into a
+// single variable for testing the resolver?
 var (
 	newTimer           = time.NewTimer
 	newTimerDNSResRate = time.NewTimer

--- a/internal/resolver/dns/dns_resolver.go
+++ b/internal/resolver/dns/dns_resolver.go
@@ -48,8 +48,10 @@ var EnableSRVLookups = false
 var logger = grpclog.Component("dns")
 
 // Globals to stub out in tests.
-var newTimer = time.NewTimer
-var newTimerDNSResRate = time.NewTimer
+var (
+	newTimer = time.NewTimer
+	newTimerDNSResRate = time.NewTimer
+)
 
 func init() {
 	resolver.Register(NewBuilder())

--- a/internal/resolver/dns/dns_resolver.go
+++ b/internal/resolver/dns/dns_resolver.go
@@ -47,8 +47,9 @@ var EnableSRVLookups = false
 
 var logger = grpclog.Component("dns")
 
-// A global to stub out in tests.
+// Globals to stub out in tests.
 var newTimer = time.NewTimer
+var newTimerDNSResRate = time.NewTimer
 
 func init() {
 	resolver.Register(NewBuilder())
@@ -219,7 +220,7 @@ func (d *dnsResolver) watcher() {
 			// Success resolving, wait for the next ResolveNow. However, also wait 30 seconds at the very least
 			// to prevent constantly re-resolving.
 			backoffIndex = 1
-			timer = time.NewTimer(minDNSResRate)
+			timer = newTimerDNSResRate(minDNSResRate)
 			select {
 			case <-d.ctx.Done():
 				timer.Stop()

--- a/internal/resolver/dns/dns_resolver.go
+++ b/internal/resolver/dns/dns_resolver.go
@@ -49,7 +49,7 @@ var logger = grpclog.Component("dns")
 
 // Globals to stub out in tests.
 var (
-	newTimer = time.NewTimer
+	newTimer           = time.NewTimer
 	newTimerDNSResRate = time.NewTimer
 )
 

--- a/internal/resolver/dns/dns_resolver_test.go
+++ b/internal/resolver/dns/dns_resolver_test.go
@@ -43,7 +43,7 @@ func TestMain(m *testing.M) {
 	// Set a non-zero duration only for tests which are actually testing that
 	// feature.
 	replaceDNSResRate(time.Duration(0)) // No nead to clean up since we os.Exit
-	replaceNetFunc(nil)                 // No nead to clean up since we os.Exit
+	overrideDefaultResolver(false)      // No nead to clean up since we os.Exit
 	code := m.Run()
 	os.Exit(code)
 }
@@ -125,13 +125,18 @@ func (*testResolver) LookupTXT(ctx context.Context, host string) ([]string, erro
 	return txtLookup(host)
 }
 
-func replaceNetFunc(ch chan struct{}) func() {
+// overrideDefaultResolver overrides the defaultResolver used by the code with
+// an instance of the testResolver. pushOnLookup controls whether the
+// testResolver created here pushes lookupHost events on its channel.
+func overrideDefaultResolver(pushOnLookup bool) func() {
 	oldResolver := defaultResolver
-	if ch != nil {
-		defaultResolver = &testResolver{ch: testutils.NewChannel()}
-	} else {
-		defaultResolver = &testResolver{}
+
+	var ch *testutils.Channel
+	if pushOnLookup {
+		ch = testutils.NewChannel()
 	}
+	defaultResolver = &testResolver{ch: ch}
+
 	return func() {
 		defaultResolver = oldResolver
 	}
@@ -1481,7 +1486,7 @@ func TestRateLimitedResolve(t *testing.T) {
 
 	// Create a new testResolver{} for this test because we want the exact count
 	// of the number of times the resolver was invoked.
-	nc := replaceNetFunc(make(chan struct{}))
+	nc := overrideDefaultResolver(true)
 	defer nc()
 
 	target := "foo.bar.com"
@@ -1509,8 +1514,7 @@ func TestRateLimitedResolve(t *testing.T) {
 
 	// Wait for the first resolution request to be done. This happens as part
 	// of the first iteration of the for loop in watcher().
-	_, err = tr.ch.Receive(ctx)
-	if err != nil {
+	if _, err := tr.ch.Receive(ctx); err != nil {
 		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 
@@ -1523,9 +1527,7 @@ func TestRateLimitedResolve(t *testing.T) {
 	continueCtx, continueCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
 	defer continueCancel()
 
-	_, err = tr.ch.Receive(continueCtx)
-
-	if err == nil {
+	if _, err := tr.ch.Receive(continueCtx); err == nil {
 		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
 	}
 
@@ -1541,8 +1543,7 @@ func TestRateLimitedResolve(t *testing.T) {
 	timerPointer.Reset(0)
 
 	// Now that DNS Min Res Rate timer has gone off, it should lookup again.
-	_, err = tr.ch.Receive(ctx)
-	if err != nil {
+	if _, err := tr.ch.Receive(ctx); err != nil {
 		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 
@@ -1552,8 +1553,7 @@ func TestRateLimitedResolve(t *testing.T) {
 		r.ResolveNow(resolver.ResolveNowOptions{})
 	}
 
-	_, err = tr.ch.Receive(continueCtx)
-	if err == nil {
+	if _, err = tr.ch.Receive(continueCtx); err == nil {
 		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
 	}
 
@@ -1566,8 +1566,7 @@ func TestRateLimitedResolve(t *testing.T) {
 	timerPointer.Reset(0)
 
 	// Now that DNS Min Res Rate timer has gone off, it should lookup again.
-	_, err = tr.ch.Receive(ctx)
-	if err != nil {
+	if _, err = tr.ch.Receive(ctx); err != nil {
 		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 

--- a/internal/resolver/dns/dns_resolver_test.go
+++ b/internal/resolver/dns/dns_resolver_test.go
@@ -107,12 +107,12 @@ type testResolver struct {
 	// A write to this channel is made when this resolver receives a resolution
 	// request. Tests can rely on reading from this channel to be notified about
 	// resolution requests instead of sleeping for a predefined period of time.
-	ch *testutils.Channel
+	lookupHostCh *testutils.Channel
 }
 
 func (tr *testResolver) LookupHost(ctx context.Context, host string) ([]string, error) {
-	if tr.ch != nil {
-		tr.ch.Send(struct{}{})
+	if tr.lookupHostCh != nil {
+		tr.lookupHostCh.Send(nil)
 	}
 	return hostLookup(host)
 }
@@ -131,11 +131,11 @@ func (*testResolver) LookupTXT(ctx context.Context, host string) ([]string, erro
 func overrideDefaultResolver(pushOnLookup bool) func() {
 	oldResolver := defaultResolver
 
-	var ch *testutils.Channel
+	var lookupHostCh *testutils.Channel
 	if pushOnLookup {
-		ch = testutils.NewChannel()
+		lookupHostCh = testutils.NewChannel()
 	}
-	defaultResolver = &testResolver{ch: ch}
+	defaultResolver = &testResolver{lookupHostCh: lookupHostCh}
 
 	return func() {
 		defaultResolver = oldResolver
@@ -1514,7 +1514,7 @@ func TestRateLimitedResolve(t *testing.T) {
 
 	// Wait for the first resolution request to be done. This happens as part
 	// of the first iteration of the for loop in watcher().
-	if _, err := tr.ch.Receive(ctx); err != nil {
+	if _, err := tr.lookupHostCh.Receive(ctx); err != nil {
 		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 
@@ -1527,7 +1527,7 @@ func TestRateLimitedResolve(t *testing.T) {
 	continueCtx, continueCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
 	defer continueCancel()
 
-	if _, err := tr.ch.Receive(continueCtx); err == nil {
+	if _, err := tr.lookupHostCh.Receive(continueCtx); err == nil {
 		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
 	}
 
@@ -1543,7 +1543,7 @@ func TestRateLimitedResolve(t *testing.T) {
 	timerPointer.Reset(0)
 
 	// Now that DNS Min Res Rate timer has gone off, it should lookup again.
-	if _, err := tr.ch.Receive(ctx); err != nil {
+	if _, err := tr.lookupHostCh.Receive(ctx); err != nil {
 		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 
@@ -1553,7 +1553,7 @@ func TestRateLimitedResolve(t *testing.T) {
 		r.ResolveNow(resolver.ResolveNowOptions{})
 	}
 
-	if _, err = tr.ch.Receive(continueCtx); err == nil {
+	if _, err = tr.lookupHostCh.Receive(continueCtx); err == nil {
 		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
 	}
 
@@ -1566,7 +1566,7 @@ func TestRateLimitedResolve(t *testing.T) {
 	timerPointer.Reset(0)
 
 	// Now that DNS Min Res Rate timer has gone off, it should lookup again.
-	if _, err = tr.ch.Receive(ctx); err != nil {
+	if _, err = tr.lookupHostCh.Receive(ctx); err != nil {
 		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 

--- a/internal/resolver/dns/dns_resolver_test.go
+++ b/internal/resolver/dns/dns_resolver_test.go
@@ -1529,6 +1529,10 @@ func TestRateLimitedResolve(t *testing.T) {
 		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
 	}
 
+	// Make the DNSMinResRate timer fire immediately (by receiving it, then
+	// resetting to 0), this will unblock the resolver which is currently
+	// blocked on the DNS Min Res Rate timer going off, which will allow it to
+	// continue to the next iteration of the watcher loop.
 	timer, err := timerChan.Receive(ctx)
 	if err != nil {
 		t.Fatalf("Error receiving timer from mock NewTimer call: %v", err)
@@ -1553,6 +1557,7 @@ func TestRateLimitedResolve(t *testing.T) {
 		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
 	}
 
+	// Make the DNSMinResRate timer fire immediately again.
 	timer, err = timerChan.Receive(ctx)
 	if err != nil {
 		t.Fatalf("Error receiving timer from mock NewTimer call: %v", err)

--- a/internal/resolver/dns/dns_resolver_test.go
+++ b/internal/resolver/dns/dns_resolver_test.go
@@ -49,8 +49,8 @@ func TestMain(m *testing.M) {
 }
 
 const (
-	txtBytesLimit      = 255
-	defaultTestTimeout = 10 * time.Second
+	txtBytesLimit           = 255
+	defaultTestTimeout      = 10 * time.Second
 	defaultTestShortTimeout = 10 * time.Millisecond
 )
 
@@ -107,12 +107,12 @@ type testResolver struct {
 	// A write to this channel is made when this resolver receives a resolution
 	// request. Tests can rely on reading from this channel to be notified about
 	// resolution requests instead of sleeping for a predefined period of time.
-	ch chan struct{}
+	ch *testutils.Channel
 }
 
 func (tr *testResolver) LookupHost(ctx context.Context, host string) ([]string, error) {
 	if tr.ch != nil {
-		tr.ch <- struct{}{}
+		tr.ch.Send(struct{}{})
 	}
 	return hostLookup(host)
 }
@@ -127,8 +127,11 @@ func (*testResolver) LookupTXT(ctx context.Context, host string) ([]string, erro
 
 func replaceNetFunc(ch chan struct{}) func() {
 	oldResolver := defaultResolver
-	defaultResolver = &testResolver{ch: ch}
-
+	if ch != nil {
+		defaultResolver = &testResolver{ch: testutils.NewChannel()}
+	} else {
+		defaultResolver = &testResolver{}
+	}
 	return func() {
 		defaultResolver = oldResolver
 	}
@@ -1452,13 +1455,15 @@ func TestCustomAuthority(t *testing.T) {
 // requests. It sets the re-resolution rate to a small value and repeatedly
 // calls ResolveNow() and ensures only the expected number of resolution
 // requests are made.
+
 func TestRateLimitedResolve(t *testing.T) {
 	defer leakcheck.Check(t)
 	defer func(nt func(d time.Duration) *time.Timer) {
 		newTimer = nt
 	}(newTimer)
 	newTimer = func(d time.Duration) *time.Timer {
-		// Will never fire on its own, will protect from triggering exponential backoff.
+		// Will never fire on its own, will protect from triggering exponential
+		// backoff.
 		return time.NewTimer(time.Hour)
 	}
 	defer func(nt func(d time.Duration) *time.Timer) {
@@ -1467,7 +1472,8 @@ func TestRateLimitedResolve(t *testing.T) {
 
 	timerChan := testutils.NewChannel()
 	newTimerDNSResRate = func(d time.Duration) *time.Timer {
-		// Will never fire on its own, allows this test to call timer immediately.
+		// Will never fire on its own, allows this test to call timer
+		// immediately.
 		t := time.NewTimer(time.Hour)
 		timerChan.Send(t)
 		return t
@@ -1498,30 +1504,29 @@ func TestRateLimitedResolve(t *testing.T) {
 		t.Fatalf("delegate resolver returned unexpected type: %T\n", tr)
 	}
 
-	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer ctxCancel()
-
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
 
 	// Wait for the first resolution request to be done. This happens as part
 	// of the first iteration of the for loop in watcher().
-	select {
-	case <-tr.ch:
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for lookup() call.")
+	_, err = tr.ch.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 
-	// Call Resolve Now 100 times, shouldn't continue onto next iteration of watcher, thus shouldn't lookup again.
+	// Call Resolve Now 100 times, shouldn't continue onto next iteration of
+	// watcher, thus shouldn't lookup again.
 	for i := 0; i <= 100; i++ {
 		r.ResolveNow(resolver.ResolveNowOptions{})
 	}
 
-	shouldNotHappenCtx, shouldNotHappenCtxCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
-	defer shouldNotHappenCtxCancel()
+	continueCtx, continueCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer continueCancel()
 
-	select {
-	case <-tr.ch:
+	_, err = tr.ch.Receive(continueCtx)
+
+	if err == nil {
 		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
-	case <-shouldNotHappenCtx.Done():
 	}
 
 	timer, err := timerChan.Receive(ctx)
@@ -1532,15 +1537,20 @@ func TestRateLimitedResolve(t *testing.T) {
 	timerPointer.Reset(0)
 
 	// Now that DNS Min Res Rate timer has gone off, it should lookup again.
-	select {
-	case <-tr.ch:
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for lookup() call.")
+	_, err = tr.ch.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 
-	// Resolve Now 1000 more times, shouldn't lookup again as DNS Min Res Rate timer has not gone off.
+	// Resolve Now 1000 more times, shouldn't lookup again as DNS Min Res Rate
+	// timer has not gone off.
 	for i := 0; i < 1000; i++ {
 		r.ResolveNow(resolver.ResolveNowOptions{})
+	}
+
+	_, err = tr.ch.Receive(continueCtx)
+	if err == nil {
+		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
 	}
 
 	timer, err = timerChan.Receive(ctx)
@@ -1551,10 +1561,9 @@ func TestRateLimitedResolve(t *testing.T) {
 	timerPointer.Reset(0)
 
 	// Now that DNS Min Res Rate timer has gone off, it should lookup again.
-	select {
-	case <-tr.ch:
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for lookup() call.")
+	_, err = tr.ch.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Timed out waiting for lookup() call.")
 	}
 
 	wantAddrs := []resolver.Address{{Addr: "1.2.3.4" + colonDefaultPort}, {Addr: "5.6.7.8" + colonDefaultPort}}

--- a/internal/resolver/dns/dns_resolver_test.go
+++ b/internal/resolver/dns/dns_resolver_test.go
@@ -51,6 +51,7 @@ func TestMain(m *testing.M) {
 const (
 	txtBytesLimit      = 255
 	defaultTestTimeout = 10 * time.Second
+	defaultTestShortTimeout = 10 * time.Millisecond
 )
 
 type testClientConn struct {
@@ -1460,10 +1461,17 @@ func TestRateLimitedResolve(t *testing.T) {
 		// Will never fire on its own, will protect from triggering exponential backoff.
 		return time.NewTimer(time.Hour)
 	}
+	defer func(nt func(d time.Duration) *time.Timer) {
+		newTimerDNSResRate = nt
+	}(newTimerDNSResRate)
 
-	const dnsResRate = 10 * time.Millisecond
-	dc := replaceDNSResRate(dnsResRate)
-	defer dc()
+	timerChan := testutils.NewChannel()
+	newTimerDNSResRate = func(d time.Duration) *time.Timer {
+		// Will never fire on its own, allows this test to call timer immediately.
+		t := time.NewTimer(time.Hour)
+		timerChan.Send(t)
+		return t
+	}
 
 	// Create a new testResolver{} for this test because we want the exact count
 	// of the number of times the resolver was invoked.
@@ -1490,55 +1498,63 @@ func TestRateLimitedResolve(t *testing.T) {
 		t.Fatalf("delegate resolver returned unexpected type: %T\n", tr)
 	}
 
-	// Observe the time before unblocking the lookupHost call.  The 100ms rate
-	// limiting timer will begin immediately after that.  This means the next
-	// resolution could happen less than 100ms if we read the time *after*
-	// receiving from tr.ch
-	start := time.Now()
+	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer ctxCancel()
+
 
 	// Wait for the first resolution request to be done. This happens as part
-	// of the first iteration of the for loop in watcher() because we call
-	// ResolveNow in Build.
-	<-tr.ch
-
-	// Here we start a couple of goroutines. One repeatedly calls ResolveNow()
-	// until asked to stop, and the other waits for two resolution requests to be
-	// made to our testResolver and stops the former. We measure the start and
-	// end times, and expect the duration elapsed to be in the interval
-	// {wantCalls*dnsResRate, wantCalls*dnsResRate}
-	done := make(chan struct{})
-	go func() {
-		for {
-			select {
-			case <-done:
-				return
-			default:
-				r.ResolveNow(resolver.ResolveNowOptions{})
-				time.Sleep(1 * time.Millisecond)
-			}
-		}
-	}()
-
-	gotCalls := 0
-	const wantCalls = 3
-	min, max := wantCalls*dnsResRate, (wantCalls+1)*dnsResRate
-	tMax := time.NewTimer(max)
-	for gotCalls != wantCalls {
-		select {
-		case <-tr.ch:
-			gotCalls++
-		case <-tMax.C:
-			t.Fatalf("Timed out waiting for %v calls after %v; got %v", wantCalls, max, gotCalls)
-		}
+	// of the first iteration of the for loop in watcher().
+	select {
+	case <-tr.ch:
+	case <-ctx.Done():
+		t.Fatal("Timed out waiting for lookup() call.")
 	}
-	close(done)
-	elapsed := time.Since(start)
 
-	if gotCalls != wantCalls {
-		t.Fatalf("resolve count mismatch for target: %q = %+v, want %+v\n", target, gotCalls, wantCalls)
+	// Call Resolve Now 100 times, shouldn't continue onto next iteration of watcher, thus shouldn't lookup again.
+	for i := 0; i <= 100; i++ {
+		r.ResolveNow(resolver.ResolveNowOptions{})
 	}
-	if elapsed < min {
-		t.Fatalf("elapsed time: %v, wanted it to be between {%v and %v}", elapsed, min, max)
+
+	shouldNotHappenCtx, shouldNotHappenCtxCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer shouldNotHappenCtxCancel()
+
+	select {
+	case <-tr.ch:
+		t.Fatalf("Should not have looked up again as DNS Min Res Rate timer has not gone off.")
+	case <-shouldNotHappenCtx.Done():
+	}
+
+	timer, err := timerChan.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Error receiving timer from mock NewTimer call: %v", err)
+	}
+	timerPointer := timer.(*time.Timer)
+	timerPointer.Reset(0)
+
+	// Now that DNS Min Res Rate timer has gone off, it should lookup again.
+	select {
+	case <-tr.ch:
+	case <-ctx.Done():
+		t.Fatal("Timed out waiting for lookup() call.")
+	}
+
+	// Resolve Now 1000 more times, shouldn't lookup again as DNS Min Res Rate timer has not gone off.
+	for i := 0; i < 1000; i++ {
+		r.ResolveNow(resolver.ResolveNowOptions{})
+	}
+
+	timer, err = timerChan.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Error receiving timer from mock NewTimer call: %v", err)
+	}
+	timerPointer = timer.(*time.Timer)
+	timerPointer.Reset(0)
+
+	// Now that DNS Min Res Rate timer has gone off, it should lookup again.
+	select {
+	case <-tr.ch:
+	case <-ctx.Done():
+		t.Fatal("Timed out waiting for lookup() call.")
 	}
 
 	wantAddrs := []resolver.Address{{Addr: "1.2.3.4" + colonDefaultPort}, {Addr: "5.6.7.8" + colonDefaultPort}}

